state-chain/pallets/cf-elections/src/electoral_systems/state_machine/consensus.rs

```diff
@@ -0,0 +1,233 @@
+// majority consensus
+
+/// Simple implementation of a (super-)majority consensus, in case of ties the last element is
+/// chosen
+pub struct SupermajorityConsensus<Vote: PartialEq> {
+	votes: BTreeMap<Vote, u32>,
+}
+
+pub struct SuccessThreshold {
+	pub success_threshold: u32,
+}
+
+impl<Vote: PartialEq> Default for SupermajorityConsensus<Vote> {
+	fn default() -> Self {
+		Self { votes: Default::default() }
+	}
+}
+
+impl<Vote: Ord + PartialEq + Clone> ConsensusMechanism for SupermajorityConsensus<Vote> {
+	type Vote = Vote;
+	type Result = Vote;
+	type Settings = SuccessThreshold;
+
+	fn insert_vote(&mut self, vote: Self::Vote) {
+		*self.votes.entry(vote).or_insert(0) += 1;
+	}
+
+	fn check_consensus(&self, settings: &Self::Settings) -> Option<Self::Result> {
+		self.votes.iter().max_by_key(|(_, count)| *count).and_then(|(vote, count)| {
+			if *count >= settings.success_threshold {
+				Some(vote.clone())
+			} else {
+				None
+			}
+		})
+	}
+}
+
+//-----------------------------------------------
+// staged consensus
+
+/// Staged consensus.
+///
+/// Votes are indexed by stages, and each stage is evaluated
+/// separately. Evaluation happens from highest priority to lowest,
+/// i.e. the highest priority stage which achieves consensus determines the result.
+/// If no stage achieves consensus, the result is inconclusive.
+pub struct StagedConsensus<Stage: ConsensusMechanism, Priority: Ord> {
+	stages: BTreeMap<Priority, Stage>,
+}
+
+pub struct StagedVote<Stage: ConsensusMechanism, Priority: Ord> {
+	pub priority: Priority,
+	pub vote: Stage::Vote,
+}
+
+impl<Stage: ConsensusMechanism, Priority: Ord> From<(Priority, Stage::Vote)>
+	for StagedVote<Stage, Priority>
+{
+	fn from((priority, vote): (Priority, Stage::Vote)) -> Self {
+		Self { priority, vote }
+	}
+}
+
+impl<Stage: ConsensusMechanism, Priority: Ord> StagedConsensus<Stage, Priority> {
+	pub fn new() -> Self {
+		Self { stages: BTreeMap::new() }
+	}
+}
+
+impl<Stage: ConsensusMechanism, Priority: Ord> Default for StagedConsensus<Stage, Priority> {
+	fn default() -> Self {
+		Self { stages: Default::default() }
+	}
+}
+
+impl<Stage: ConsensusMechanism, Priority: Ord + Copy> ConsensusMechanism
+	for StagedConsensus<Stage, Priority>
+{
+	type Result = Stage::Result;
+	type Vote = StagedVote<Stage, Priority>;
+	type Settings = Stage::Settings;
+
+	fn insert_vote(&mut self, vote: Self::Vote) {
+		self.stages.entry(vote.priority).or_default().insert_vote(vote.vote);
+	}
+
+	fn check_consensus(&self, settings: &Self::Settings) -> Option<Self::Result> {
+		// we check all stages starting with the highest index,
+		// the first one that has consensus wins
+		for (_, stage) in self.stages.iter().rev() {
+			if let Some(result) = stage.check_consensus(settings) {
+				return Some(result);
+			}
+		}
+
+		None
+	}
+}
+
+//------ multiple votes -----------
+/// This is a consensus modifier which allows multiple votes to be cast
+pub struct MultipleVotes<Base: ConsensusMechanism> {
```

Comment:
As discussed - this doesn't protect against multiple voting, which is fine in the context in which it's used. But we should add a comment mentioning this.

Reply:
Done.